### PR TITLE
JDK-8318444: Write details about compilation bailouts into crash reports

### PR DESCRIPTION
**Pull request by tstuefe**
A little debugging aid to help analyze broken bailout chains, mainly in C2 (C1 is pretty clean).

A broken bailout chain occurs when code marks a compilation as failed, but then either that function itself or any of its caller functions fails to abort the compilation. That may cause crashes, e.g. [JDK-8318183](https://bugs.openjdk.org/browse/JDK-8318183) or [JDK-8318445](https://bugs.openjdk.org/browse/JDK-8318445).

Now, if the compiler initiates a bailout, it stores some context information - compile id, time, and call stack. That information is stored as part of `Compile` or `Compilation`, depending on the compiler.

If we crash later during the same compilation, we print out that information as part of the crash report. That way, we have two call stacks, and it is easy to spot where the compiler failed to heed the bailout.

---------

Looks like this (from https://github.com/openjdk/jdk/pull/16248). The first call stack is the crash point. The second call stack is the point where the compiler bailout was initiated. 

```
Current CompileTask:
C2:2574 45 45  843       4       sun.nio.fs.UnixPath::resolve (17 bytes)

Stack: [0x00007fa608cb3000,0x00007fa608db4000],  sp=0x00007fa608daf310,  free space=1008k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x631bb4]  Unique_Node_List::push(Node*)+0x20  (node.hpp:1650)
V  [libjvm.so+0xb8ea65]  ConnectionGraph::verify_ram_nodes(Compile*, Node*)+0x87  (escape.cpp:743)
V  [libjvm.so+0x960dda]  Compile::Optimize()+0x956  (compile.cpp:2361)
V  [libjvm.so+0x959d6c]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x165e  (compile.cpp:860)
V  [libjvm.so+0x81bcd9]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x203  (c2compiler.cpp:134)
V  [libjvm.so+0x97bf63]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0xac5  (compileBroker.cpp:2290)
V  [libjvm.so+0x97a981]  CompileBroker::compiler_thread_loop()+0x411  (compileBroker.cpp:1951)
V  [libjvm.so+0x99ebc0]  CompilerThread::thread_entry(JavaThread*, JavaThread*)+0x84  (compilerThread.cpp:61)
V  [libjvm.so+0xde0050]  JavaThread::thread_main_inner()+0x15c  (javaThread.cpp:720)
V  [libjvm.so+0xddfeea]  JavaThread::run()+0x258  (javaThread.cpp:705)
V  [libjvm.so+0x15f5a04]  Thread::call_run()+0x1a8  (thread.cpp:220)
V  [libjvm.so+0x12de0a2]  thread_native_entry(Thread*)+0x1c3  (os_linux.cpp:785)

siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x0000000000000028

Pending compilation failure details for thread 0x00007fa5b8003140:
  Time: 2.545243 seconds (0d 0h 0m 2s)  Compile id: 843
  Reason: 'hit memory limit while compiling'
  Callstack:[0x00007fa64262bd64] Compile::record_method_not_compilable(char const*)+0x4a
[0x00007fa6428be4d5] Compile::record_method_not_compilable_oom()+0x27
[0x00007fa6427c7130] Compile::check_node_count(unsigned int, char const*)+0x34
[0x00007fa64329bb06] PhaseIterGVN::optimize()+0xaa
[0x00007fa6428b2d75] Compile::Optimize()+0x8f1
[0x00007fa6428abd6c] Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x165e
[0x00007fa64276dcd9] C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x203
[0x00007fa6428cdf63] CompileBroker::invoke_compiler_on_method(CompileTask*)+0xac5
[0x00007fa6428cc981] CompileBroker::compiler_thread_loop()+0x411
[0x00007fa6428f0bc0] CompilerThread::thread_entry(JavaThread*, JavaThread*)+0x84
[0x00007fa642d32050] JavaThread::thread_main_inner()+0x15c
[0x00007fa642d31eea] JavaThread::run()+0x258
[0x00007fa643547a04] Thread::call_run()+0x1a8
[0x00007fa6432300a2] thread_native_entry(Thread*)+0x1c3

```

/label remove hotspot
/label add hotspot-compiler
/reviewers 2

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8318444](https://bugs.openjdk.org/browse/JDK-8318444): Write details about compilation bailouts into crash reports (**Enhancement** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [e9d6830b](https://git.openjdk.org/jdk/pull/16247/files/e9d6830b230eac805b1c5e7a0b333729c28ad3d5)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16247/head:pull/16247` \
`$ git checkout pull/16247`

Update a local copy of the PR: \
`$ git checkout pull/16247` \
`$ git pull https://git.openjdk.org/jdk.git pull/16247/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16247`

View PR using the GUI difftool: \
`$ git pr show -t 16247`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16247.diff">https://git.openjdk.org/jdk/pull/16247.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16247#issuecomment-1770270009)

### CODE REVIEW

File: src/hotspot/share/utilities/vmError.cpp
Comment (TobiHartmann) [inline src/hotspot/share/utilities/vmError.cpp:1054]:
```suggestion
          _verbose && _thread != nullptr && _thread->is_Compiler_thread())
```

File: src/hotspot/share/compiler/compilationFailureInfo.hpp
Comment (TobiHartmann) [inline src/hotspot/share/compiler/compilationFailureInfo.hpp:49]:
```suggestion
  // current thread is compiler thread and there is an ongoing compilation
```

**Review by TobiHartmann — APPROVED** [review]
That looks good to me but someone more familiar with error reporting should have a look as well. I submitted some testing and will report back once it passed.

**Comment by tstuefe** [comment]
> That looks good to me but someone more familiar with error reporting should have a look as well. I submitted some testing and will report back once it passed.

Thank you @TobiHartmann ! I'll look around for some runtime folks to give a second review.

**Comment by TobiHartmann** [comment]
All tests passed.

**Comment by tstuefe** [comment]
> All tests passed.

Thank you Tobias.

**Comment by tstuefe** [comment]
@dholmes-ora If you got time, could you possibly look at this? I need the runtime changes okayed. 

In summary, the changes in os.cpp/hpp just split out a new function "print_elapsed_time", and the changes in error handling print the attached compiler information into the hs-err file.

**Comment by tstuefe** [comment]
May I have a second reviewer? @chhagedorn maybe? You seem to be uniquely suited for compiler/error reporting reviews :)

File: src/hotspot/share/compiler/compilationFailureInfo.cpp
Comment (chhagedorn) [inline src/hotspot/share/compiler/compilationFailureInfo.cpp:90]: I suggest to use braces for the ifs.

File: src/hotspot/share/compiler/compilationFailureInfo.cpp
Comment (chhagedorn) [inline src/hotspot/share/compiler/compilationFailureInfo.cpp:45]: Is there a specific reason to define this separately and not directly use `_stack(2)` below?

File: src/hotspot/share/opto/compile.cpp
Comment (chhagedorn) [inline src/hotspot/share/opto/compile.cpp:995]: Should we first check for `nullptr` before deleting here?

**Review by chhagedorn — APPROVED** [review]
That's a nice addition! I have some small comments but otherwise, looks good.

File: src/hotspot/share/opto/compile.cpp
Comment (tstuefe) [reply src/hotspot/share/opto/compile.cpp:995]: delete, like free, accepts nullptr and does nothing. We delete/os::free null in many places.

**Comment by tstuefe** [comment]
Many thanks, @chhagedorn !

Will wait for GHAs to complete, then push.

File: src/hotspot/share/compiler/compilationFailureInfo.cpp
Comment (chhagedorn) [reply src/hotspot/share/compiler/compilationFailureInfo.cpp]: Hehe ok. I guess with the IDE parameter hints, it's probably fine to pass a 2 directly. But either way is fine :-)

**Review by chhagedorn — APPROVED** [review]
Thanks for the updates, still looks good!

**Pull request closed without merging**
